### PR TITLE
GMMK 2 WBG7 MCU compatibility 

### PR DESCRIPTION
**Pull request by GloriousThrall**
## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [ ] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by wb-Joy** [comment]
Can anyone review this PR？

File: keyboards/gmmk/gmmk2/config.h [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "config_common.h"
+
+#define RGB_DISABLE_WHEN_USB_SUSPENDED TRUE
```
Comment (drashna) [inline keyboards/gmmk/gmmk2/config.h:21]:
```suggestion
#define RGB_DISABLE_WHEN_USB_SUSPENDED
```

File: keyboards/gmmk/gmmk2/config.h [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+#define ENABLE_RGB_MATRIX_SOLID_REACTIVE_MULTICROSS
+#define ENABLE_RGB_MATRIX_SOLID_REACTIVE_NEXUS
+#define ENABLE_RGB_MATRIX_SOLID_REACTIVE_MULTINEXUS
+#define ENABLE_RGB_MATRIX_SPLASH
+#define ENABLE_RGB_MATRIX_MULTISPLASH
+#define ENABLE_RGB_MATRIX_SOLID_SPLASH
+#define ENABLE_RGB_MATRIX_SOLID_MULTISPLASH
+
+/* 1000Hz USB polling - it's the default on stock firmware */
+#define USB_POLLING_INTERVAL_MS 1
```
Comment (drashna) [inline keyboards/gmmk/gmmk2/config.h:81]:
```suggestion
```

File: keyboards/gmmk/gmmk2/p65/ansi/keymaps/via/config.h [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#define KEYBOARD_SHARED_EP TRUE
```
Comment (drashna) [inline keyboards/gmmk/gmmk2/p65/ansi/keymaps/via/config.h:17]: Belongs in rules.mk, not here.

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/gmmk/gmmk2/p65/ansi/keymaps/via/config.h (hunk 3 above)
Comment (wb-Joy) [reply keyboards/gmmk/gmmk2/p65/ansi/keymaps/via/config.h]: Has been resolved.

File: keyboards/gmmk/gmmk2/p65/ansi/keymaps/via/config.h [hunk 4]
```diff
@@ -0,0 +1 @@
+#define STM32_USB_USE_OTG1 TRUE
```
Comment (drashna) [inline keyboards/gmmk/gmmk2/p65/ansi/keymaps/via/config.h:1]: THis needs a license header and an include guard (`#pragma once`).

File: keyboards/gmmk/gmmk2/p65/ansi/rules.mk [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+CONSOLE_ENABLE = no          # Console for debug.
+COMMAND_ENABLE = no          # Commands for debug and configuration.
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no        # Breathing sleep LED during USB suspend.
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes            # USB Nkey Rollover.
+BACKLIGHT_ENABLE = no        # Enable keyboard backlight functionality.
+RGBLIGHT_ENABLE = no         # Enable keyboard RGB underglow.
+BLUETOOTH_ENABLE = no        # Enable Bluetooth.
+AUDIO_ENABLE = no            # Audio output.
```
Comment (drashna) [inline keyboards/gmmk/gmmk2/p65/ansi/rules.mk:22]:
```suggestion
NKRO_ENABLE = yes            # Enable NKRO Rollover.
BACKLIGHT_ENABLE = no        # Enable keyboard backlight functionality.
RGBLIGHT_ENABLE = no         # Enable keyboard RGB underglow.
AUDIO_ENABLE = no            # Audio output.
```

File: keyboards/gmmk/gmmk2/p65/iso/rules.mk [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+CONSOLE_ENABLE = no          # Console for debug.
+COMMAND_ENABLE = no          # Commands for debug and configuration.
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no        # Breathing sleep LED during USB suspend.
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes            # USB Nkey Rollover.
+BACKLIGHT_ENABLE = no        # Enable keyboard backlight functionality.
+RGBLIGHT_ENABLE = no         # Enable keyboard RGB underglow.
+BLUETOOTH_ENABLE = no        # Enable Bluetooth.
+AUDIO_ENABLE = no            # Audio output.
```
Comment (drashna) [inline keyboards/gmmk/gmmk2/p65/iso/rules.mk:22]:
```suggestion
NKRO_ENABLE = yes            # Enable NKRO Rollover.
BACKLIGHT_ENABLE = no        # Enable keyboard backlight functionality.
RGBLIGHT_ENABLE = no         # Enable keyboard RGB underglow.
AUDIO_ENABLE = no            # Audio output.
```

File: keyboards/gmmk/gmmk2/p96/ansi/keymaps/default/keymap.c [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+
+// Each layer gets a name for readability, which is then used in the keymap matrix below.
+// The underscores don't mean anything - you can have a layer called STUFF or any other name.
+#define _BL 0
+#define _FL 1
```
Comment (drashna) [inline keyboards/gmmk/gmmk2/p96/ansi/keymaps/default/keymap.c:22]:
```suggestion
enum custom_layers {
    _BL,
    _FL,
};
```

File: keyboards/gmmk/gmmk2/p96/ansi/keymaps/via/config.h [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#define DYNAMIC_KEYMAP_LAYER_COUNT 3
```
Comment (drashna) [inline keyboards/gmmk/gmmk2/p96/ansi/keymaps/via/config.h:17]:
```suggestion
#pragma once

#define DYNAMIC_KEYMAP_LAYER_COUNT 3
```

File: keyboards/gmmk/gmmk2/p96/ansi/keymaps/via/keymap.c [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+
+// Each layer gets a name for readability, which is then used in the keymap matrix below.
+// The underscores don't mean anything - you can have a layer called STUFF or any other name.
+#define _BL 0
+#define _FL 1
```
Comment (drashna) [inline keyboards/gmmk/gmmk2/p96/ansi/keymaps/via/keymap.c:22]:
```suggestion
enum custom_layers {
    _BL,
    _FL,
};
```

File: keyboards/gmmk/gmmk2/p96/ansi/rules.mk [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+EXTRAKEY_ENABLE = yes        # Audio control and System control.
+CONSOLE_ENABLE = no          # Console for debug.
+COMMAND_ENABLE = no          # Commands for debug and configuration.
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no        # Breathing sleep LED during USB suspend.
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes            # USB Nkey Rollover.
+BACKLIGHT_ENABLE = no        # Enable keyboard backlight functionality.
+RGBLIGHT_ENABLE = no         # Enable keyboard RGB underglow.
+BLUETOOTH_ENABLE = no        # Enable Bluetooth.
```
Comment (drashna) [inline keyboards/gmmk/gmmk2/p96/ansi/rules.mk:21]:
```suggestion
NKRO_ENABLE = yes            # Enable NKRO Rollover.
BACKLIGHT_ENABLE = no        # Enable keyboard backlight functionality.
RGBLIGHT_ENABLE = no         # Enable keyboard RGB underglow.
```

File: keyboards/gmmk/gmmk2/p96/iso/keymaps/via/config.h [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#define DYNAMIC_KEYMAP_LAYER_COUNT 3
```
Comment (drashna) [inline keyboards/gmmk/gmmk2/p96/iso/keymaps/via/config.h:17]:
```suggestion
#pragma once

#define DYNAMIC_KEYMAP_LAYER_COUNT 3
```

File: keyboards/gmmk/gmmk2/p96/iso/rules.mk [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+CONSOLE_ENABLE = no          # Console for debug.
+COMMAND_ENABLE = no          # Commands for debug and configuration.
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no        # Breathing sleep LED during USB suspend.
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes            # USB Nkey Rollover.
+BACKLIGHT_ENABLE = no        # Enable keyboard backlight functionality.
+RGBLIGHT_ENABLE = no         # Enable keyboard RGB underglow.
+BLUETOOTH_ENABLE = no        # Enable Bluetooth.
+AUDIO_ENABLE = no            # Audio output.
```
Comment (drashna) [inline keyboards/gmmk/gmmk2/p96/iso/rules.mk:22]:
```suggestion
NKRO_ENABLE = yes            # Enable NKRO Rollover.
BACKLIGHT_ENABLE = no        # Enable keyboard backlight functionality.
RGBLIGHT_ENABLE = no         # Enable keyboard RGB underglow.
AUDIO_ENABLE = no            # Audio output.
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/gmmk/gmmk2/p65/ansi/keymaps/via/config.h (hunk 4 above)
Comment (wb-Joy) [reply keyboards/gmmk/gmmk2/p65/ansi/keymaps/via/config.h:19]: resolved.

**Comment by drashna** [comment]
Honestly, I'm not exactly thrilled about having a custom spi_master.c, nor the flash stuff handled in the keyboard level, like this. 

but aside from those, otherwise looks good

**Comment by wb-Joy** [comment]
> Honestly, I'm not exactly thrilled about having a custom spi_master.c, nor the flash stuff handled in the keyboard level, like this.
> 
> but aside from those, otherwise looks good

This is because the hardware was designed with two SPI masters, now PCBA can not be modified, so it is implemented in this way.

**Comment by perehonchuk** [comment]
hello, any updates about this pr?
It would be great to have qmk support for gmmk2

**Review by drashna — APPROVED** [review]

File: keyboards/gmmk/gmmk2/p96/ansi/rules.mk [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+COMMAND_ENABLE = no          # Commands for debug and configuration.
+NKRO_ENABLE = yes            # Enable NKRO Rollover.
+BACKLIGHT_ENABLE = no        # Enable keyboard backlight functionality.
+RGBLIGHT_ENABLE = no         # Enable keyboard RGB underglow.
+AUDIO_ENABLE = no            # Audio output.
+RGB_MATRIX_ENABLE = yes      # Enable RGB matrix effects.
+RGB_MATRIX_DRIVER = AW20216  # Enable RGB matrix effects.
+FLASH_DRIVER = spi           # Enable SPI flash
+EEPROM_DRIVER = custom       # Enable SPI flash memory
+SRC +=  src/eeprom_flash.c
```
Comment (tzarc) [inline keyboards/gmmk/gmmk2/p96/ansi/rules.mk:23]:
```suggestion
EEPROM_DRIVER = wear_leveling
WEAR_LEVELING_DRIVER = spi_flash
```

File: keyboards/gmmk/gmmk2/p96/src/eeprom_flash.c [hunk 14]
```diff
@@ -0,0 +1,692 @@
+/* Copyright (C) 2021 Westberry Technology (ChangZhou) Corp., Ltd
```
Comment (tzarc) [inline keyboards/gmmk/gmmk2/p96/src/eeprom_flash.c:1]: Please remove this file, it has been made redundant by the wear-leveling system.

**Review by tzarc — CHANGES_REQUESTED** [review]

**Comment by joshualex** [comment]
Trying to compile with the PR branch I now get:
```
EEPROM_DRIVER=wear_leveling is not a valid EEPROM driver
```

**Comment by KarlK90** [comment]
This has to be rebased on the current qmk develop branch

**Comment by joshualex** [comment]
@KarlK90 this has not been merged yet... you mean rebase it?

**Comment by KarlK90** [comment]
Yes, the wear leveling for SPI flash was only merged 5 days ago into develop

**Comment by KarlK90** [comment]
Also `make gmmk/gmmk2/p96/ansi:default` does error out during linking:

```
Linking: .build/gmmk_gmmk2_p96_ansi_default.elf                                                     [ERRORS]
 | 
 | /usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld:rules_memory.ld:314 cannot move location counter backwards (from 0000000020009680 to 0000000020007000)
 | collect2: error: ld returned 1 exit status
 | 
make[1]: *** [builddefs/common_rules.mk:267: .build/gmmk_gmmk2_p96_ansi_default.elf] Error 1
Make finished with errors
make: *** [Makefile:414: gmmk/gmmk2/p96/ansi:default] Error 1
```

Looks like it consumes to much ram now.

**Comment by wb-Joy** [comment]
@joshualex @KarlK90 Now you can clone the gmmk_V2_keyboard branch directly. I've solved the RAM overflow problem and merged the develop branch.

**Review by tzarc — APPROVED** [review]

**Comment by guilleortas** [comment]
@GloriousThrall What about the GMMK 2 65% ?
Thanks!

**Comment by itarze** [comment]
#18185 This PR. @guilleortas

**Comment by guilleortas** [comment]
I was able to build and flash QMK with VIA support for my GMMK 2 65% ISO. However, it shows up as GMMK Pro in VIA, which has a different layout and nothing matches. Would this be solved using a via.json design file? If so, could you provide it?
Thanks!

**Pull request merged**